### PR TITLE
Fix update_ancestor_stakes in locktower

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Locktower vote threshold checks are busted b/c we are not calculating the stakes with the correct set of parents

#### Summary of Changes
Generate the stakes for the parents of the last unsimulated vote in collect_vote_lockouts()

Fixes #
https://github.com/solana-labs/solana/issues/3625

### CODE REVIEW

File: core/src/locktower.rs [hunk 1]
```diff
@@ -162,8 +162,17 @@ impl Locktower {
                 };
                 Self::update_ancestor_lockouts(&mut stake_lockouts, &vote, ancestors);
             }
-            // each account hash a stake for all the forks in the active tree for this bank
-            Self::update_ancestor_stakes(&mut stake_lockouts, bank_slot, lamports, ancestors);
+
+            if vote_state.votes.len() > 1 {
+                let last_unsimulated_vote = &vote_state.votes[vote_state.votes.len() - 2];
```
Comment (sagar-solana) [inline core/src/locktower.rs]: nit: last_real_vote or something not negative :)

Comment (aeyakovenko) [reply core/src/locktower.rs]: There is a VoteState method for this, nth_recent_vote

File: core/src/locktower.rs [hunk 2]
```diff
@@ -162,8 +162,17 @@ impl Locktower {
                 };
                 Self::update_ancestor_lockouts(&mut stake_lockouts, &vote, ancestors);
             }
-            // each account hash a stake for all the forks in the active tree for this bank
-            Self::update_ancestor_stakes(&mut stake_lockouts, bank_slot, lamports, ancestors);
+
+            if vote_state.votes.len() > 1 {
+                let last_real_vote = &vote_state.votes[vote_state.votes.len() - 2];
```
Comment (garious) [inline core/src/locktower.rs]: That 2 seems arbitrary to me. Are you planning to add a test before this goes in?

Comment (aeyakovenko) [reply core/src/locktower.rs]: should be `nth_recent_vote(1)`

**Comment by aeyakovenko** [comment]
needs tests

**Comment by carllin** [comment]
yup, writing them now.

File: core/src/locktower.rs (hunk 2 above)
Comment (aeyakovenko) [reply core/src/locktower.rs]: can you add a comment explaining that `0` is the simulated vote.

Comment (carllin) [reply core/src/locktower.rs]: sure thing

File: core/src/locktower.rs [hunk 3]
```diff
@@ -162,8 +162,19 @@ impl Locktower {
                 };
                 Self::update_ancestor_lockouts(&mut stake_lockouts, &vote, ancestors);
             }
-            // each account hash a stake for all the forks in the active tree for this bank
-            Self::update_ancestor_stakes(&mut stake_lockouts, bank_slot, lamports, ancestors);
+
+            if vote_state.votes.len() > 1 {
```
Comment (aeyakovenko) [inline core/src/locktower.rs]: Is the empty check necessary since nth returns an option?

Comment (carllin) [reply core/src/locktower.rs]: nope not necessary, removed in the latest

File: core/src/locktower.rs [hunk 4]
```diff
@@ -728,4 +744,55 @@ mod test {
...
+        let stakes: HashMap<_, _> = accounts.iter().map(|(pk, a)| (*pk, a.lamports)).collect();
+        let epoch_stakes = EpochStakes::new(0, stakes, &Pubkey::default());
+        let mut locktower = Locktower::new(epoch_stakes, VOTE_THRESHOLD_DEPTH, threshold_size);
+
+        // CASE 1: Record the first VOTE_THRESHOLD locktower votes for fork 2. We want to
+        // evaluate a vote on slot VOTE_THRESHOLD_DEPTH. The nth most recent vote should be
+        // for slot 0, which is common to all account vote states, so we should pass the
+        // threshold check
+        let vote_to_evaluate = VOTE_THRESHOLD_DEPTH as u64;
+        for vote in locktower_votes[..locktower_votes.len()].iter() {
```
Comment (garious) [inline core/src/locktower.rs]: What's going on here? That looks equivalent to `& locktower_votes`

Comment (carllin) [reply core/src/locktower.rs]: haha yeah they are exactly equivalent, I had it as ...locktower_votes.len() - 1 at one point, and forgot to adjust it afterwards. Good catch!

**Pull request merged**